File: code/JavaLab/src/main/java/algorithm/leetcode/linklist/206.反转链表.java
package algorithm.leetcode.linklist;

import algorithm.config.ListNode;

/**
 * https://leetcode-cn.com/problems/reverse-linked-list/
 * <p>
 * 翻转链表
 */
class LC206 {

    // 非递归写法
    public ListNode reverseList(ListNode head) {
        if (head == null) {
            return null;
        }
        ListNode p = head;
        ListNode pre = null;
        while (p != null) {
            ListNode next = p.next;
            p.next = pre;
            pre = p;
            p = next;
        }
        return pre;
//        return reverse(null, head);
    }


    // 递归写法
    public ListNode reverse(ListNode pre, ListNode cur) {
        if (cur == null) {
            return pre;
        }
        ListNode next = cur.next;
        cur.next = pre;
        return reverse(cur, next);
    }

}

File: code/ahasmove/206.反转链表.cpp
/*
 * @lc app=leetcode.cn id=206 lang=cpp
 *
 * [206] 反转链表
 *
 * https://leetcode-cn.com/problems/reverse-linked-list/description/
 *
 * algorithms
 * Easy (58.31%)
 * Likes:    442
 * Dislikes: 0
 * Total Accepted:    61.1K
 * Total Submissions: 99K
 * Testcase Example:  '[1,2,3,4,5]'
 *
 * 反转一个单链表。
 * 
 * 示例:
 * 
 * 输入: 1->2->3->4->5->NULL
 * 输出: 5->4->3->2->1->NULL
 * 
 * 进阶:
 * 你可以迭代或递归地反转链表。你能否用两种方法解决这道题？
 * 
 */
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution
{
public:
    ListNode *reverseList(ListNode *head)
    {
        
        if(head == nullptr)
            return head;
        
        ListNode *pre = nullptr;
        while (head->next != nullptr)
        {
            ListNode *cur = head;
            head = head->next;
            cur->next = pre;
            pre = cur;
        }
        head->next = pre;
        return head;
    }
};
